valve-monitor/web-app/src/services/prisma-datastore/CompositeDataProvider.ts

```diff
@@ -33,6 +37,102 @@ export default class CompositeDataProvider implements DataProvider {
+    // get as much device data as is available in Prisma
+    const valveMonitorDevices =
+      await this.prismaDataProvider.getValveMonitorDeviceData();
+
+    // fetch any env vars for each device
+    const deviceEnvVars = await Promise.all(
+      valveMonitorDevices.map((device) =>
+        this.getDeviceEnvVars(device.deviceID)
+      )
+    );
+
+    // filter out any devices not assigned to fleets
+    const devicesWithFleets = valveMonitorDevices.filter(
+      (device) => device.deviceFleetID !== undefined
+    );
+
+    // fetch fleet env vars
+    let fleetEnvVars: FleetEnvVars[];
+    if (devicesWithFleets.length > 0) {
+      fleetEnvVars = await Promise.all(
+        devicesWithFleets.map((device) =>
+          this.getFleetEnvVars(device.deviceFleetID)
+        )
+      );
+    }
+
+    // combine devices with any env vars
+    const normalizedDeviceData = valveMonitorDevices.map((device) => {
+      const filteredDeviceEnvVars = deviceEnvVars.filter(
+        (deviceEnvVar) => deviceEnvVar.deviceID === device.deviceID
+      );
+
+      // extract any device env vars
+      const deviceMonitorInterval =
+        filteredDeviceEnvVars[0].environment_variables.monitor_interval;
+
+      const deviceMinFlowThreshold =
+        filteredDeviceEnvVars[0].environment_variables
+          .flow_rate_alarm_threshold_min;
+
+      const deviceMaxFlowThreshold =
+        filteredDeviceEnvVars[0].environment_variables
+          .flow_rate_alarm_threshold_max;
+
+      const filteredFleetEnvVars = uniqBy(fleetEnvVars, "fleetUID").filter(
+        (fleetEnvVar) => fleetEnvVar.fleetUID === device.deviceFleetID
+      );
+
+      // extract any fleet env vars
+      const fleetMonitorInterval =
+        filteredFleetEnvVars[0] &&
+        filteredFleetEnvVars[0].environment_variables.monitor_interval
+          ? filteredFleetEnvVars[0].environment_variables.monitor_interval
+          : undefined;
+
+      const fleetMinFlowThreshold =
+        filteredFleetEnvVars[0] &&
+        filteredFleetEnvVars[0].environment_variables
+          .flow_rate_alarm_threshold_min
+          ? filteredFleetEnvVars[0].environment_variables
+              .flow_rate_alarm_threshold_min
+          : undefined;
+
+      const fleetMaxFlowThreshold =
+        filteredFleetEnvVars[0] &&
+        filteredFleetEnvVars[0].environment_variables
+          .flow_rate_alarm_threshold_max
+          ? filteredFleetEnvVars[0].environment_variables
+              .flow_rate_alarm_threshold_max
+          : undefined;
+
+      // device env vars trump fleet env vars
+      const monitorFrequency =
+        deviceMonitorInterval || fleetMonitorInterval || "x";
+
+      const minFlowThreshold =
+        deviceMinFlowThreshold || fleetMinFlowThreshold || "xx.x";
+
+      const maxFlowThreshold =
+        deviceMaxFlowThreshold || fleetMaxFlowThreshold || "xx.x";
+
+      const formattedDeviceObj = {
+        ...device,
+        deviceAlarm: !device.deviceAlarm ? `-` : `!`,
```

Comment:
I’d flip this check as it’d let you get rid of the leading `!` which I think makes this easier to read.

Reply:
Flipped